### ISSUE TITLE
Redis cache hash support possible?

### ISSUE THREAD

**Issue by andrewpros** [yiisoft/yii2-redis#111]
Hi.
Is there, a possible way to support redis hash in cache?
The use case is simple, caching large list of data is not optimal when you just need to get random key from that list multiple times in request life,  the way it works now forces getting the whole list and pick one or more item, but redis has specialized hash type for such scenario that is designed for this.

I mean i know that this would need changes in yii2 cache, so i would say lets treat this as a loose discussion.

It is the first time i came across this problem and it occurred to me, looks like no one is interested in extending the cache system in yii even if it is oblivious.

If you look at redis you clearly see that there is possibility that some software can have specialized options depending on the data type to store/cache.

But the cache system in yii2 doesn't have any way to set additional options that could be used by the implementation to use that specialized features of target software depending on data type, there is only key, value, duration.

And i also know that other frameworks don't support this either, which surprises me even more.

Basically no way to set additional parameters with set, get etc. (like optional $options parameter)

I probably will end up with my own implementation of redis cache that will give me access to that functionality out of the box (will need some dirty hax), but i'm curious what do you think about it?

And i'm sure there are plenty more software than redis with specialized options for data types beyond key, value and duration not possible to use out of the box in basically every php framework cache system.

Or i get this wrong and this is fine?

**Comment by cebe** [comment]
If you have configured `Yii::$app->cache` to be `yii\redis\Cache`, you can do the following:

```php
Yii::$app->cache->redis->hset('mykey', 'somefield', 'somevalue');
Yii::$app->cache->redis->hget('mykey', 'somefield');
...
```

All redis methods are available in this way: https://github.com/yiisoft/yii2-redis/blob/c50382e4d9d10d13d2c998baf973d68e06fa6977/Connection.php#L27-L224

**Comment by andrewpros** [comment]
Thank you for answer, but this isn't exactly what i had in mind, this is not using the cache layer of Yii2, it just calls redis commands and i can call them directly by redis component, and i know this, actually there is also no point in using it like that, via cache component access, Yii::$app->redis->hset is the same.

And yes i know that redis is a cache itself, but i was specifically discussing the cache system possibilities.

As this is not using cache system there is no out of the box expire or dependency support nor higher level cache functions like getOrSet.

Which again means, custom implementation needs to be written and if i really want to use the standardized yii2 cache interface i need to make some dirty hacks. But as i stated, popular other frameworks are also lacking such support.

**Comment by cebe** [comment]
I don't really get what you are asking for. The cache interface implements what's common about different cache systems. If you need redis specific features you need to call redis methods.

**Comment by andrewpros** [comment]
Well i was just pointing out that i was asking about cache and that its implementation is lacking in yii and other frameworks, but your example is not about cache anymore, just direct use of redis component.

I'm just surprised that something such obvious as an ability to set additional cache options is not present anywhere, making it impossible to use all functionalities of target software out of the box. Especially when it is so simple to support this.

Thats it.

I also think that the documentation should be updated, because if you use HMGET and simply pass a list of string delimited fields (like you do on redis) or array of fields as the parameter expecting this to work, well, it will not work, returned data will be empty without any error.

Thats because the code only supports variable length arguments and if you use it accessing component Yii::$app->redis you won't see the phpdoc about this.

With HMGET (or any multi field command) it is probably the most used and expected scenario just to pass an array of fields and thus it would be good to note this in docs that it needs to be unpacked to argument list.
